functions/snippets/src/main/java/com/example/functions/OcrProcessImage.java

```
@@ -0,0 +1,152 @@
+  private static final Gson gson = new Gson();
+
+  public OcrProcessImage() throws IOException {
+    publisher = Publisher.newBuilder(
+        ProjectTopicName.of(PROJECT_ID, TRANSLATE_TOPIC_NAME)).build();
+  }
+
+  @Override
+  public void accept(GcsEvent gcsEvent, Context context) throws RuntimeException {
+
+    // Validate parameters
+    String bucket = gcsEvent.getBucket();
+    if (bucket == null) {
+      throw new IllegalArgumentException("Missing bucket parameter");
+    }
+    String filename = gcsEvent.getName();
+    if (filename == null) {
+      throw new IllegalArgumentException("Missing name parameter");
+    }
+
+    detectText(bucket, filename);
+  }
+  // [END functions_ocr_process]
+
+  // [START functions_ocr_detect]
+  private void detectText(String bucket, String filename) throws RuntimeException {
+    LOGGER.info("Looking for text in image " + filename);
+
+    List<AnnotateImageRequest> visionRequests = new ArrayList<>();
+    String gcsPath = String.format("gs://%s/%s", bucket, filename);
+
+    ImageSource imgSource = ImageSource.newBuilder().setGcsImageUri(gcsPath).build();
+    Image img = Image.newBuilder().setSource(imgSource).build();
+
+    Feature textFeature = Feature.newBuilder().setType(Feature.Type.TEXT_DETECTION).build();
+    AnnotateImageRequest visionRequest =
+        AnnotateImageRequest.newBuilder().addFeatures(textFeature).setImage(img).build();
+    visionRequests.add(visionRequest);
+
+    // Detect text in an image using the Cloud Vision API
+    AnnotateImageResponse visionResponse;
+    try (ImageAnnotatorClient client = ImageAnnotatorClient.create()) {
+      visionResponse = client.batchAnnotateImages(visionRequests).getResponses(0);
+      if (visionResponse.hasError()) {
+        LOGGER.severe("Error detecting text: " + visionResponse.getError().getMessage());
+        return;
+      }
+    } catch (IOException e) {
+      // Cast to RuntimeException
+      throw new RuntimeException(e);
+    }
```

Why do we do this?

---

Just because you use a try w/ resources doesn't mean you need to ever catch anything.

---

The Base `BackgroundFunction` type in the Functions Framework can **only** throw subclasses of `RuntimeException`, which these are not.

(IntelliJ requires the exception to be addressed in a `catch` block - even if that block merely ignores the error.)

---

https://google.github.io/styleguide/javaguide.html#s6.2-caught-exceptions
Also see Effective Java v3 Exceptions
* Item 69: Use exceptions only for exceptional conditions 
* Item 70: Use checked exceptions for recoverable conditions and runtime exceptions for programming errors
* Item 71: Avoid unnecessary use of checked exceptions 
* Item 72: Favor the use of standard exceptions 
* Item 73: Throw exceptions appropriate to the abstraction
* Item 74: Document all exceptions thrown by each method 
* Item 77: Don’t ignore exceptions

---

What if you just say `throws IOException` ?

---

Ah - so (some of) these seem more like "checked" exceptions.

Would it be better to log them and exit the function?

---

Please log over `RuntimeException` everywhere.